test/jdk/jdk/jfr/event/io/TestIOTopFrame.java

```diff
@@ -0,0 +1,336 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.jfr.event.io;
+
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.channels.AsynchronousFileChannel;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import jdk.jfr.Recording;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.jfr.consumer.RecordedMethod;
+import jdk.jfr.consumer.RecordedStackTrace;
+import jdk.jfr.consumer.RecordingFile;
+
+/**
+ * @test TestIOTopFrame
+ * @summary Tests that the top frames of I/O events are as expected.
+ * @requires vm.flagless
+ * @requires vm.hasJFR
+ * @library /test/lib /test/jdk
+ * @run main/othervm jdk.jfr.event.io.TestIOTopFrame
+ */
+
+// Lines commented with a number indicate that they are the nth event for that
+// top frame. The invocation of the methods RandomAccessFile::readUTF(),
+// SocketInputStream::readNBytes(...) and FileInputStream::readAllBytes()
+// results in 2-3 events.
+public class TestIOTopFrame {
+    private static final String EVENT_FILE_READ = "jdk.FileRead";
+    private static final String EVENT_FILE_FORCE = "jdk.FileForce";
+    private static final String EVENT_FILE_WRITE = "jdk.FileWrite";
+    private static final String EVENT_SOCKET_READ = "jdk.SocketRead";
+    private static final String EVENT_SOCKET_WRITE = "jdk.SocketWrite";
+
+    public static void main(String... args) throws Exception {
+        testFileRead();
+        testFileWrite();
+        testSocketStreams();
+        testSocketChannels();
+    }
+
+    private static void testFileRead() throws Exception {
```

Comment:
In addition to java.io, there will be FileRead/FileWrite events recorded for the stream/channels created with Files.newXXX methods.